### PR TITLE
Add public posts index

### PR DESCRIPTION
**Pull request by dNitza**
This PR adds a publicly accessible index of posts at `/posts`

Todo:
- [x] Add links to post detail page
- [x] Decide whether we want to add an extra field for the post "summary" or whether we want to generate it dynamically.
- [x] Pagination
- [x] Format post date

Have left the newer/older pagination links as they're greyed out in the design if we're at the start/end of a list. It might be better to just make them text without the anchor rather than try and disable a link.

### CODE REVIEW

File: apps/admin/lib/admin/persistence/repositories/posts.rb [hunk 1]
```diff
@@ -21,8 +21,13 @@ def slug_exists?(slug)
...
         end
 
-        def listing
-          posts.order(:created_at).reverse.as(Entities::Post).to_a
+        def listing(per_page: 20, page: 1)
+          posts
+            .per_page(per_page)
+            .page(page)
+            .order(:created_at)
+            .reverse
```
Comment (timriley) [inline apps/admin/lib/admin/persistence/repositories/posts.rb]: Just checking: this is actually a modification to the SQL statement, right? And not an in-memory array reversal?

File: apps/main/lib/main/entities/post.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    class Post < Dry::Types::Value
+      Status = Types::Strict::String.default("draft").enum("draft", "published", "deleted")
+
+      attribute :id, Types::Strict::Int
+      attribute :title, Types::Strict::String
+      attribute :body, Types::String
+      attribute :teaser, Types::String
+      attribute :slug, Types::Strict::String
+      attribute :status, Status
+      attribute :author_id, Types::Int
```
Comment (timriley) [inline apps/main/lib/main/entities/post.rb]: I think we'd want `Types::Strict::Int` here?

File: apps/main/lib/main/entities/post.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      attribute :published_at, Types::DateTime
```
Comment (timriley) [inline apps/main/lib/main/entities/post.rb]: I think we'd want `Types::Strict::Time` here? AFAIK it's better to work with plain `Time` objects if we can: https://gist.github.com/pixeltrix/e2298822dd89d854444b

File: lib/persistence/relations/posts.rb [hunk 4]
```diff
@@ -4,13 +4,20 @@ class Posts < ROM::Relation[:sql]
       schema(:posts) do
         attribute :id, Types::Serial
         attribute :title, Types::String
+        attribute :teaser, Types::String
         attribute :body, Types::String
         attribute :slug, Types::String
         attribute :status, Types::String
-        attribute :author_id, Types::Serial
+        attribute :author_id, Types::ForeignKey(:users)
```
Comment (timriley) [inline lib/persistence/relations/posts.rb]: So nice :)

File: apps/main/lib/main/entities/post.rb (hunk 2 above)
Comment (dNitza) [inline apps/main/lib/main/entities/post.rb]: Yep, good spotting. If it were left as `Types::Int` would that allow an integer as a string?

File: apps/admin/lib/admin/persistence/repositories/posts.rb (hunk 1 above)
Comment (dNitza) [inline apps/admin/lib/admin/persistence/repositories/posts.rb]: Yeah that looks like an in-memory array reversal. We'd probably want something like `.order(Sequel.desc(:published_at))`?

Comment (timriley) [inline apps/admin/lib/admin/persistence/repositories/posts.rb]: Yeah, that makes sense.

File: apps/admin/lib/admin/entities/post.rb [hunk 5]
```diff
@@ -7,6 +7,7 @@ class Post < Dry::Types::Value
 
       attribute :id, Types::Strict::Int
       attribute :title, Types::Strict::String
+      attribute :teaser, Types::String
```
Comment (timriley) [inline apps/admin/lib/admin/entities/post.rb]: should be `Types::Strict::String`?

Comment (timriley) [inline apps/admin/lib/admin/entities/post.rb]: Same for body below, I guess :)

File: apps/admin/lib/admin/pager.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+      end
+    end
+
+    def prev_page
+      if current_page > 1
+        current_page - 1
+      end
+    end
+  end
+end
```
Comment (timriley) [inline apps/admin/lib/admin/pager.rb]: I'm not sure we need this class until we want to paginate on results that aren't coming out of a ROM query. I believe Andrew added this in ALS for paginating over Elasticsearch results.

File: apps/admin/lib/admin/views/posts/index.rb [hunk 7]
```diff
@@ -12,8 +13,12 @@ class Index < Admin::View
         end
 
         def locals(options = {})
+          options = { per_page: 20, page: 1 }.merge(options)
+          all_posts = posts.listing(page: options[:page], per_page: options[:per_page])
+
           super.merge(
-            posts: posts.listing
+            posts: all_posts.to_a,
```
Comment (timriley) [inline apps/admin/lib/admin/views/posts/index.rb]: Do we actually need to `#to_a` this?

File: apps/admin/lib/admin/pager.rb (hunk 6 above)
Comment (dNitza) [inline apps/admin/lib/admin/pager.rb]: Oh cool! I will use ROM's pagination then :)

File: apps/admin/lib/admin/views/posts/index.rb (hunk 7 above)
Comment (dNitza) [inline apps/admin/lib/admin/views/posts/index.rb]: Nope, the listing method already returns an array

File: apps/admin/web/templates/posts/index.html.slim [hunk 8]
```diff
@@ -8,24 +8,31 @@ h1.h-1 Posts
...
         tbody
           - posts.each do |post|
             tr id="post-#{post.slug}"
-              td.f-mono = post.id
               td = post.title
-              td = post.body
-              td.f-mono = post.slug
-              td = post.status
+              td.vertical-align--middle
+                - if post.status == "published"
```
Comment (timriley) [inline apps/admin/web/templates/posts/index.html.slim]: Wonder if we should make `post.published?` methods on the entity to make this cleaner?

Comment (timriley) [inline apps/admin/web/templates/posts/index.html.slim]: Actually, this is the admin, probably less important to do that.

What we should probably do is add a decorator and have a `#status_indicator_level` method, that returns "warning" "success" or "ghost" :)

That way we can drop this whole if/elsif/else altogether, right?

File: apps/main/lib/main/entities/post.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+module Main
+  module Entities
+    class Post < Dry::Types::Value
+      Status = Types::Strict::String.default("draft").enum("draft", "published", "deleted")
+
+      attribute :id, Types::Strict::Int
+      attribute :title, Types::Strict::String
+      attribute :body, Types::String
+      attribute :teaser, Types::String
```
Comment (timriley) [inline apps/main/lib/main/entities/post.rb]: `Strict::String`?

File: apps/main/lib/main/persistence/repositories/posts.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+          posts.by_id(id).as(Entities::Post).one
+        end
+
+        def by_slug(slug)
+          posts.by_slug(slug).as(Entities::Post).one
+        end
+
+        def slug_exists?(slug)
+          !!posts.matching_slugs(slug).one
+        end
```
Comment (timriley) [inline apps/main/lib/main/persistence/repositories/posts.rb]: Is this needed in the main repo? I thought it was only for slug uniqueness validations?

File: apps/main/lib/main/persistence/repositories/posts.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+module Main
+  module Persistence
+    module Repositories
+      class Posts < Berg::Repository[:posts]
+        relations :posts, :users
+
+        def [](id)
+          posts.by_id(id).as(Entities::Post).one
+        end
```
Comment (timriley) [inline apps/main/lib/main/persistence/repositories/posts.rb]: Is this needed in the main repo?

File: apps/admin/web/templates/posts/index.html.slim (hunk 8 above)
Comment (dNitza) [inline apps/admin/web/templates/posts/index.html.slim]: Yeah I think that's a good idea since we reuse `post.status == "published"` when checking if the detail view should be publicly accessible in another PR.

File: apps/main/lib/main/persistence/repositories/posts.rb (hunk 10 above)
Comment (dNitza) [inline apps/main/lib/main/persistence/repositories/posts.rb]: Ack! No neither of those are needed. I forgot to remove them.

File: spec/admin/shared/app/admin_posts.rb [hunk 12]
```diff
@@ -1,10 +1,14 @@
+require "babosa"
```
Comment (timriley) [inline spec/admin/shared/app/admin_posts.rb]: It's not clear why this `require` is here. Is it an indication that we're not requiring it in some implementation source file that actually uses it?

Comment (timriley) [inline spec/admin/shared/app/admin_posts.rb]: Oh, never mind, saw the `#to_slug` 😩

**Comment by timriley** [comment]
This looks pretty good on the whole. Are you able to add an integration test, @dNitza?

**Comment by dNitza** [comment]
Yep, for sure @timriley

**Comment by dNitza** [comment]
I think this is ready to merge now.

**Comment by timriley** [comment]
@dNitza Please go ahead!

**Comment by dNitza** [comment]
Manually merged :)
